kernel/conversion.ml

```diff
@@ -208,28 +222,28 @@ let convert_constructors_gen cmp_instances cmp_cumul (mind, ind, cns) nargs u1 u
   | Some _ ->
     let num_cnstr_args = constructor_cumulativity_arguments (mind,ind,cns) in
     if not (Int.equal num_cnstr_args nargs) then
-      if UVars.Instance.equal u1 u2 then s else raise MustExpand
+      if UVars.Instance.equal u1 u2 then Result.Ok s else raise MustExpand
     else
       (** By invariant, both constructors have a common supertype,
           so they are convertible _at that type_. *)
       (* NB: no variance for qualities *)
       let variance = Array.make (snd (UVars.Instance.length u1)) UVars.Variance.Irrelevant in
       cmp_cumul CONV variance u1 u2 s
 
-let convert_constructors ctor nargs u1 u2 (s, check) =
-  convert_constructors_gen (check.compare_instances ~flex:false) check.compare_cumul_instances
-    ctor nargs u1 u2 s, check
+let convert_constructors infos ctor nargs u1 u2 (s, check) =
+  fail_check infos @@ (convert_constructors_gen (check.compare_instances ~flex:false) check.compare_cumul_instances
```

Comment:
fail_check is done here but eg for the similar convert_instances it's done by the caller
Probably we should do either all fail_check inside the subfunction or all fail_check in the caller (eqappr)

also what's the point of using @@ with a parenthesized argument?

Reply:
> Probably we should do either all fail_check inside the subfunction

The problem is that some functions are exported in the API and some not, so I'd have to duplicate the code. Conversely, calling fail_check in the caller would be a bit too heavy syntactically IMO.

Reply:
Make internal _exn versions of the functions then? eg convert_constructors_exn = fail_check @@ convert_constructors

Reply:
I wend with the caller-side conversion, it's not so heavy after all.